ppa/archive/models.py

```diff
@@ -12,15 +12,38 @@ class Collection(models.Model):
     def __str__(self):
         return self.name
 
+    def full_index(self, params=None):
+        '''Index or reindex a collection's associated works.'''
+        # guard against this being called on an unsaved instance
+        if not self.pk:
+            raise ValueError(
+                'Collection instance needs to have a primary key value before '
+                'this method is called.'
+            )
+        solr, solr_collection = get_solr_connection()
+        digworks = self.digitizedwork_set.all()
+        solr_docs = [work.index_data() for work in digworks]
+        # adapted from hathi import logic
+        solr.index(solr_collection, solr_docs,
+            params=params)
+
     def save(self, *args, **kwargs):
         '''Override method so that on save, any associated
-        works have their names updated.'''
-        # Note: This is a potentially expensive operation,
-        # but collection shouldn't change often.
-        super(Collection, self).save(*args, **kwargs)
-        digworks = self.digitizedwork_set.all()
-        for work in digworks:
-            work.index()
+        works have their names updated if collection name has changed.'''
+        # first handle cases where this is a new save, just save and return
+        if not self.pk:
+            super(Collection, self).save(*args, **kwargs)
+            return
+        # if it has been saved, get the original
+        orig = Collection.objects.get(pk=self.pk)
```

Comment:
Might make sense to store original name when the object is first initialized so we don't have to make another db call (there are examples of that on Stack Overflow), but this is fine for now.